### PR TITLE
e2e: Switch to testnet fixture

### PR DESCRIPTION
**Pull request by marun**
## Why this should be merged

This change updates the e2e test suite to use the new testnet fixture introduced in #1700. The new fixture is simpler and maintained in tree to better enable e2e test development.

As part of this change, the fixture's funded keys are allocated at-most-once to tests to avoid the following problems:
 - reusing the same pre-defined set of funded addresses (e.g. the EWOQ key) precludes proper test isolation
 - iterating against a persistent network with a limited set of fixed addresses is more likely to result in failure due to funds exhaustion
 - the wallet's lack of concurrency precludes tests using a common set of keys from executing in parallel 

## How this works

- The use of ANR in the e2e suite is replaced by the testnet fixture
- Tests are updated to retrieve funded keys at runtime
- Keys are served via a simple http server to allow at-most-once allocation even when ginkgo is running specs in parallel
  - ginkgo's [strategy for parallel execution](https://onsi.github.io/ginkgo/#spec-parallelization) precludes a shared-state (i.e. non-ipc) solution

## How this was tested

- `./scripts/tests.e2e.sh` was updated to run tests in parallel and in random order. This revealed the requirement to tag the virtuous tests as `Serial` to avoid skipping on conflict with other tests.
- `./scripts/tests.e2e.persistent.sh` was updated to run `tests.e2e.sh` against a persistent network

## TODO

- [x] Merge the parent PR (#1700)

### CODE REVIEW

File: tests/fixture/test_data_server_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			keys, err := AllocateFundedKeys(uri, tc.count)
+			if tc.expectedAddresses == nil {
+				require.ErrorIs(err, errRequestedKeyCountExceedsAvailable)
+			} else {
+				require.NoError(err)
+				addresses := []string{}
+				for _, key := range keys {
+					addresses = append(addresses, key.Address().String())
+				}
+				require.Equal(tc.expectedAddresses, addresses)
```
Comment (darioush) [inline tests/fixture/test_data_server_test.go:66]: nit: Are the calls to `.String()` needed for this test?

**Review by darioush — APPROVED** [review]

File: tests/e2e/p/permissionless_subnets.go [hunk 2]
```diff
@@ -199,7 +202,7 @@ var _ = e2e.DescribePChain("[Permissionless Subnets]", func() {
 				_, err := pWallet.IssueAddPermissionlessDelegatorTx(
 					&txs.SubnetValidator{
 						Validator: txs.Validator{
-							NodeID: genesis.LocalConfig.InitialStakers[0].NodeID,
+							NodeID: validatorID,
```
Comment (hexfusion) [inline tests/e2e/p/permissionless_subnets.go:205]: :heart:

File: tests/fixture/test_data_server.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+type TestData struct {
+	FundedKeys []*secp256k1.PrivateKey
+}
+
+// http server allocating resources to tests potentially executing in parallel
+type testDataServer struct {
+	TestData
+
+	// Synchronizes access to test data
+	mutex sync.Mutex
```
Comment (hexfusion) [inline tests/fixture/test_data_server.go:42]: minor standardization style nit
```suggestion
	lock sync.Mutex
```

File: tests/fixture/test_data_server.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	return address, nil
+}
+
+// Retrieve the specified number of funded test keys from the provided URI. A given
+// key is allocated at most once during the life of the test data server.
+func AllocateFundedKeys(uri string, count int) ([]*secp256k1.PrivateKey, error) {
+	if count <= 0 {
+		return nil, errInvalidKeyCount
+	}
+	query := fmt.Sprintf("%s%s?%s=%d", uri, allocateKeysPath, keyCountParameterName, count)
```
Comment (hexfusion) [inline tests/fixture/test_data_server.go:128]: would be nice if this was using "net/url" or something else :)

File: tests/fixture/test_data_server.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		s.FundedKeys[i] = nil
+	}
+	s.FundedKeys = s.FundedKeys[:len(s.FundedKeys)-keyCount]
+}
+
+// Serve test data via http to ensure allocation is synchronized even when
+// ginkgo specs are executing in parallel. Returns the URI to access the server.
+func ServeTestData(testData TestData) (string, error) {
+	// Listen on a dynamic port to avoid conflicting with other applications
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
```
Comment (hexfusion) [inline tests/fixture/test_data_server.go:96]: don't we need to eventually close this listener?

File: tests/e2e/README.md [hunk 6]
```diff
@@ -29,8 +28,51 @@ Create a new package to implement feature-specific tests, or add tests to an exi
...
+
+# Start a new test run using the persistent network
+ginkgo -v ./tests/e2e -- \
+    --avalanchego-path=/path/to/avalanchego \
+    --ginkgo-focus-file=[name of file containing test] \
+    --use-persistent-network \
+    --network-dir=/path/to/network
+
+# Also possible to set the AVALANCHEGO_PATH env var instead of supplying --avalanchego-path
+# Also possible to set the TESTNETCTL_NETWORK_DIR env var instead of supplying --network-dir
```
Comment (hexfusion) [inline tests/e2e/README.md:75]:
```suggestion
# or TESTNETCTL_NETWORK_DIR instead of supplying --network-dir
```

**Review by hexfusion — APPROVED** [review]
Looks good few minor nits

File: tests/fixture/testnet/local/network.go [hunk 7]
```diff
@@ -57,6 +57,7 @@ func FindNextNetworkID(rootDir string) (uint32, string, error) {
 	for {
 		_, reserved := constants.NetworkIDToNetworkName[networkID]
 		if reserved {
+			networkID++
```
Comment (StephenButtolph) [inline tests/fixture/testnet/local/network.go:60]: Nice catch

File: tests/e2e/x/transfer/virtuous.go [hunk 8]
```diff
@@ -33,29 +33,48 @@ const (
...
+				for _, metrics := range allNodeMetrics {
+					if metrics[metricBlksProcessing] > 0 {
+						return false
+					}
+				}
+				return true
+			}).
+				WithTimeout(e2e.DefaultTimeout).
+				WithPolling(e2e.DefaultPollingInterval).
+				Should(gomega.BeTrue(), "The cluster is generating ongoing blocks. Is this test being run in parallel?")
```
Comment (StephenButtolph) [inline tests/e2e/x/transfer/virtuous.go:64]: I spent like 20m understanding how this works lol.

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**